tools/pythonpkg/src/python_conversion.cpp

```diff
@@ -0,0 +1,178 @@
+
+#include "datetime.h" //From Python
+
+namespace duckdb {
+
+bool DictionaryHasMapFormat(py::handle dict_values, idx_t len) {
+	if (len != 2) {
+		return false;
+	}
+	auto keys = dict_values.attr("__getitem__")(0);
+	auto values = dict_values.attr("__getitem__")(1);
+	// Dont check for 'py::list' to allow ducktyping
+	if (!py::hasattr(keys, "__getitem__") || !py::hasattr(keys, "__len__")) {
+		// throw std::runtime_error("Dictionary malformed, keys(index 0) found within 'dict.values' is not a list");
+		return false;
+	}
+	if (!py::hasattr(values, "__getitem__") || !py::hasattr(values, "__len__")) {
+		// throw std::runtime_error("Dictionary malformed, values(index 1) found within 'dict.values' is not a list");
+		return false;
+	}
+	auto size = py::len(keys);
+	if (size != py::len(values)) {
+		// throw std::runtime_error("Dictionary malformed, keys and values lists are not of the same size");
+		return false;
+	}
+	return true;
+}
+
+Value TransformDictionaryToMap(py::handle dict_values) {
+	auto keys = dict_values.attr("__getitem__")(0);
+	auto values = dict_values.attr("__getitem__")(1);
+
+	auto key_size = py::len(keys);
+	D_ASSERT(key_size == py::len(values));
+	if (key_size == 0) {
+		// dict == { 'key': [], 'value': [] }
+		return Value::MAP(Value::EMPTYLIST(LogicalType::SQLNULL), Value::EMPTYLIST(LogicalType::SQLNULL));
+	}
+	auto key_list = TransformPythonValue(keys);
+	auto value_list = TransformPythonValue(values);
+	return Value::MAP(key_list, value_list);
+}
+
+static void CastAllValues(vector<Value> &values, const LogicalType &target_type) {
+	for (idx_t i = 0; i < values.size(); i++) {
+		values[i] = values[i].CastAs(target_type);
+	}
+}
+
+static void CastNullValues(vector<Value> &values, vector<idx_t> &null_indices, const LogicalType &target_type) {
+	for (auto &idx : null_indices) {
+		values[idx] = values[idx].CastAs(target_type);
+	}
+}
+
+Value TransformListValue(py::handle ele) {
+	auto size = py::len(ele);
+
+	if (size == 0) {
+		return Value::EMPTYLIST(LogicalType::SQLNULL);
+	}
+
+	vector<Value> values;
+	values.reserve(size);
+
+	LogicalType element_type = LogicalType::SQLNULL;
+	for (idx_t i = 0; i < size; i++) {
+		Value new_value = TransformPythonValue(ele.attr("__getitem__")(i));
+		element_type = LogicalType::MaxLogicalType(element_type, new_value.type());
+		values.push_back(move(new_value));
+	}
+
+	return Value::LIST(element_type, values);
+}
+
+Value TransformPythonValue(py::handle ele) {
+	auto datetime_mod = py::module::import("datetime");
+	auto datetime_date = datetime_mod.attr("date");
+	auto datetime_datetime = datetime_mod.attr("datetime");
+	auto datetime_time = datetime_mod.attr("time");
+	auto decimal_mod = py::module::import("decimal");
+	auto decimal_decimal = decimal_mod.attr("Decimal");
+	auto numpy_mod = py::module::import("numpy");
+	auto numpy_ndarray = numpy_mod.attr("ndarray");
+	auto uuid_mod = py::module::import("uuid");
+	auto uuid_uuid = uuid_mod.attr("UUID");
```

Comment:
I wonder about the performance cost of doing all of these imports once per value - can we cache these somewhere (e.g. in the connection)? Python caches the actual modules but this will still likely be 10+ dictionary lookups per value which could be very significant.

Reply:
But I guess since we know which modules+types we're looking for, we can probably do something smarter that doesn't require so many lookups

Reply:
I would go with a class that caches a fixed set of imports, e.g. something like this:

```cpp
struct PyImportCache {
  py::obj datetime_mod;
  py::obj datetime_date;
  py::obj datetime_datetime;
  py::obj datetime_time;
  py::obj decimal_mod;
  py::obj decimal_decimal;
  py::obj numpy_mod;
  py::obj numpy_ndarray;
  py::obj uuid_mod;
  py::obj uuid_uuid;
};
```

I think this is useful elsewhere in the Python API too. Then we can extend it as required with new cached imports, and no string lookups are required at all.